jenkins/test-history/gen_json.py

```diff
@@ -23,198 +23,335 @@
+            for job, build, timestamp in jobs_and_builds)
+
+    for job, build, timestamp, build_tests in builds_tests_iterator:
+        print('%s/%s' % (job, build))
+        build_info = jobs.setdefault(job, {}).setdefault(build, {})
+        build_info['timestamp'] = timestamp
+        build_info['tests'] = []
+        for name, duration, failed, skipped in build_tests:
+            result = {
+                'name': names.index(name),
+                'time': duration,
+            }
+            if failed:
+                result['failed'] = True
             if skipped:
-                continue
-            if job not in tests[name]:
-                tests[name][job] = []
-            tests[name][job].append({
-                'build': build,
-                'failed': failed,
-                'time': duration
-            })
-    return tests
+                result['skipped'] = True
+            build_info['tests'].append(result)
+
 
+def remove_old_builds(buckets, now):
+    pruned = 0
+    for jobs in buckets.itervalues():
+        for job_name, job in jobs.iteritems():
+            for build_num, build in job.items():  # intentional copy
+                if build['timestamp'] < now - MAX_AGE:
+                    pruned += 1
+                    job.pop(build_num)
+    print('pruned %d old builds' % pruned)
 
-def main(server, match):
+
+def main(jobs_dir, match, outfile, threads, client_class=GCSClient):
     """Collect test info in matching jobs."""
-    print('Finding tests in jobs matching {} at server {}'.format(
-        match, server))
+    print('Finding tests in jobs matching %s' % match)
     matcher = re.compile(match).match
-    tests = get_tests(server, matcher)
-    with open('tests.json', 'w') as buf:
+    tests = None
+    if os.path.exists(outfile):
+        try:
+            tests = json.load(open(outfile))
+            if 'test_names' not in tests:
+                raise ValueError
+        except ValueError:
+            tests = None
+        else:
+            print('Resuming from previous run...')
+            names = IndexedList(tests['test_names'])
+            tests['test_names'] = names
+            remove_old_builds(tests['buckets'], time.time())
+    if tests is None:
+        names = IndexedList()
+        tests = {'test_names': names, 'buckets': {}}
+    for bucket in jobs_dir.split(','):
+        if not bucket.endswith('/'):
+            bucket += '/'
+        bucket_jobs = tests['buckets'].setdefault(bucket, {})
+        get_tests(names, bucket, matcher, threads, client_class, bucket_jobs)
+    with open(outfile, 'w') as buf:
         json.dump(tests, buf, sort_keys=True)
 
 
 def get_options(argv):
     """Process command line arguments."""
     parser = argparse.ArgumentParser()
     parser.add_argument(
-        '--server',
-        help='hostname of jenkins server',
-        required=True,
+        '--jobs_dir',
+        help='comma-separated locations of test artifacts on GCS',
```

Comment:
rather than comma-separated, isn't it possible to get a list by doing `--jobs_dir` more than once?